CHAID/CHAID.py

```diff
@@ -366,37 +369,63 @@ def node(self, rows, ind, dep, depth=0, parent=None, parent_decisions=None):
                 self.node_count += 1
         return self.tree_store
 
-    def generate_best_split(self, ind, dep):
+    def generate_best_split(self, ind, dep, wt):
         """ internal method to generate the best split """
         split = CHAIDSplit(None, None, None, 1)
         relative_split_threshold = 1 - self.split_threshold
+
         for i, index in enumerate(ind):
             index = index.deep_copy()
             unique = set(index.arr)
 
             mappings = MappingDict()
-            frequencies = {}
+            freq = {}
+            wt_freq = {}
             for col in unique:
                 counts = np.unique(dep.arr[index.arr == col], return_counts=True)
-                frequencies[col] = cl.defaultdict(int)
-                frequencies[col].update(np.transpose(counts))
+                freq[col] = cl.defaultdict(int)
+                wt_freq[col] = cl.defaultdict(int)
+                freq[col].update(np.transpose(counts))
+                if wt is not None:
+                    for dep_v in set(dep.arr):
+                        wt_freq[col][dep_v] = wt[(index.arr == col) * (dep.arr == dep_v)].sum()
 
             while len(unique) > 1:
                 size = int((len(unique) * (len(unique) - 1)) / 2)
                 sub_data_columns = [('combinations', object), ('chi', float), ('p', float)]
                 sub_data = np.array([(None, 0, 1)]*size, dtype=sub_data_columns, order='F')
                 for j, comb in enumerate(it.combinations(unique, 2)):
-                    col1_freq = frequencies[comb[0]]
-                    col2_freq = frequencies[comb[1]]
+                    col1_freq = freq[comb[0]]
+                    col2_freq = freq[comb[1]]
 
                     keys = set(col1_freq.keys()).union(col2_freq.keys())
 
-                    cr_table = [
+                    n_ij = np.array([
                         [col1_freq.get(k, 0) for k in keys],
                         [col2_freq.get(k, 0) for k in keys]
-                    ]
+                    ])
+
+                    if wt is not None:
+                        col1_wt_freq = wt_freq[comb[0]]
+                        col2_wt_freq = wt_freq[comb[1]]
+
+                        m_ij = w_ij = n_ij / np.array([
+                            [col1_wt_freq.get(k, 0) for k in keys],
+                            [col2_wt_freq.get(k, 0) for k in keys]
+                        ])
+
+                        alpha, beta, eps = (1, 1, 1)
+                        while eps > 0.000001:
+                            alpha = alpha * np.vstack(n_ij.sum(axis=1) / m_ij.sum(axis=1))
+                            beta = n_ij.sum(axis=0) / (alpha * w_ij).sum(axis=0)
+                            eps = np.max(np.absolute(w_ij * alpha * beta - m_ij))
+                            m_ij = w_ij * alpha * beta
+                        n_ij = m_ij
+
+                    exp = (np.vstack(n_ij.sum(axis=1)) * n_ij.sum(axis=0)) / n_ij.sum().astype(float)
```

Comment:
I don't think we need vstack here on the assumption that the ndarray is always square. Is it safe to make that assumption?